Adding the recommended way of authentication for ManageIQ.

vi-patel opened the pull request:
It is recommended from ManageIQ that multiple REST calls should use the token for authentication.
Authentication will always use a token, authentication will work the following way:

1. If a user supplies a token, attempt to auth w/the token and if it fails, authentication will fail
2. If the user does not supply a token, check for $HOME/.miqcli/auth file
    validate that it is good,
    if not use, passed username/password to generate a new token, and update $HOME/.miqcli/auth
    else
    auth fails
3. create $HOME/.miqcli/auth
   use username/password to generate a new token, and update $HOME/.miqcli/auth

Temporary change:
updating the git location to use github's master
I couldn't use the latest released (0.3.0), as token support was recently added.
This can be update as soon as: https://github.com/ManageIQ/manageiq-api-client-python/issues/31 is resolved, to release a new version of manageiq-api-client.

This PR closes #33

vi-patel committed:
Updating the authentication using the manageiq_client api.

Authentication will occur the following way:
1. if user passes token it will attempt auth with that
2. check for an auth file w/token and validate use if it works, else
   generate a token using username/password
3. if there is no auth file, generatea a token using username/password.

vi-patel committed:
Merge remote-tracking branch 'origin' into auth

ryankwilliams commented on miqcli/api.py:
```diff
@@ -63,16 +68,79 @@ def connect(self, settings):
+                                   "{0}".format(exception))
+        # check for an auth token file & validate
+        if os.path.isfile(authfile):
+            with file(authfile) as f:
```
Could we use open instead of file? This provides Python 3.x compatibility.

```python
-            with file(authfile) as f:
+            with open(authfile) as f:
```

ryankwilliams commented on miqcli/api.py:
```diff
@@ -63,16 +68,79 @@ def connect(self, settings):
+        """
+        method to get a token if username/password is set
+        """
+        auth_endpoint = self.settings["url"] + "/auth"
```
Looks like the **settings** attribute is unresolved. This attribute is only in the scope of the Collection class, which is a child class of the parent class ClientAPI. If we want the ClientAPI class to have an attribute for the MIQ configuration settings. We should look into passing the settings to the ClientAPI class. Then the Collection class would still have a settings attribute.

```python
-    def __init__(self):
-        """Constructor."""
+    def __init__(self, settings):
+        """Constructor.
+
+        :param settings: MIQ settings
+        :type settings: dict
+        """
+        self.settings = settings
         self._client = None
```

Then within the miqcli/collections/__init__.py we could remove the settings attribute from there.

```python
-    _settings = None
-
     def __init__(self, settings):
         """
         :param settings: MIQ settings
         :type settings: dict
         """
-         super(Collection, self).__init__()
+         super(Collection, self).__init__(settings)
-        self._settings = settings
-
-    @property
-    def settings(self):
-        """
-        :return: dict of settings
-        """
-        return self._settings
+        super(Collection, self).__init__(settings)
```

Those code snippets were not tested, just put as an example.

ryankwilliams commented on miqcli/constants.py:
```diff
@@ -27,6 +27,7 @@
     'username': 'admin',
     'password': 'smartvm',
-    'url': 'https://localhost:8443',
+    'url': 'https://localhost:8443/api',
```
Users now will have to provide the keyword **/api** to their MIQ URL if they wish to override the default settings? What if we handle adding the api keyword when creating the endpoint url to generate the authentication token? This way then the user does not need to provide it. I am assuming if the user does not provide the api keyword in their URL the request call would fail above since the endpoint is not correctly set.

We can always check before adding the api keyword to the endpoint to see if the user already included it. That way we do not append it again.

What do you think?

ryankwilliams commented on miqcli/api.py:
```diff
@@ -63,16 +68,79 @@ def connect(self, settings):
+            with open(authfile, "w") as f:
+                f.write(token)
+
+            return (True, None)
```
Just a comment: the parentheses are redundant.

vi-patel committed:
Updating based on review

* setting auth url back, so /api is not needed
* moving settings property to API class instead of Collection
* cleanup

vi-patel committed:
Merge branch 'auth' of https://github.com/vi-patel/manageiq-cli into auth

Commit:
removing SHA1 from git link

According to this bug https://bugs.launchpad.net/pbr/+bug/1677243 there is a problem with SHA1 in the git link

Commit:
put back SHA1 and add pbr to latest version

ryankwilliams commented:
LGTM.

@smatula if you approve of these changes. We just will want to make sure we squash and remove the extra commits leaving one commit for the merge.
